fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/visitor/RecordQueryPlannerSubstitutionVisitor.java

```diff
@@ -102,6 +102,7 @@ public static RecordQueryPlan removeIndexFetch(@Nonnull RecordMetaData recordMet
                 // Need the primary key, even if it wasn't one of the explicit result fields.
                 fields.addAll(commonPrimaryKey.normalizeKeyForPositions());
             }
+            fields.removeIf(KeyExpression::canBeComputedWithoutMessage);
```

Comment:
What is this transformation attempting to accomplish? Is it just that you can get the record type key (and any literal or empty expressions) without needing to fetch the record, so if those are the only fields missing from the required results, then this should still allow the index plan to be used as a covering index?

I'm not sure this is totally right. I guess the other way of doing this would be to somehow tell the `AvailableFields` data structure how to resolve a `recordType` key expression. That might be necessary if this is ever generalized to work on multi-type indexes (which I see that it's not designed for at the moment).

Reply:
Fields of `RecordTypeKeyExpression` cannot be in `AvailableFields` unless we reimplement that logic in there as that logic wants to create copiers which you cannot do for this case. Also, it's really not made available in the index, but just an idiosyncrasy that some key expressions can be evaluated regardless of what the index defines, e.g. constants, record type.

Reply:
Do you have a suggested name for this property?

Reply:
I almost feel like `InferrableOnSingleType` or something might make sense. It's a little unfortunate, because the property is something like `Inferrable` (for literals and empty) or "is record type key". It also seems pretty specific to this exact circumstance, but ¯\\\_(ツ)_/¯ I guess that happens